In [2]:
import pandas as pd #data processing, CSV file I/O (e.g. pd.read_csv)
import numpy as np # numerical calculation, linear algebra
import matplotlib.pyplot as plt # visulizing plot
import seaborn as sns # for visilizing plot

In [3]:
# Lets check the columns name in dataframe as all columns are vot visible
pd.set_option('display.max_columns', None)

In [4]:
df = pd.read_csv("DataCoSupplyChainDataset.csv", encoding="latin1")
df.head() # Showing first 5 rows in dataframe, we can set df.head(number_of_rows) where as number_of_rows for how many rows we want to see

Type  Days for shipping (real)  Days for shipment (scheduled)  \
0     DEBIT                         3                              4   
1  TRANSFER                         5                              4   
2      CASH                         4                              4   
3     DEBIT                         3                              4   
4   PAYMENT                         2                              4   

   Benefit per order  Sales per customer   Delivery Status  \
0          91.250000          314.640015  Advance shipping   
1        -249.089996          311.359985     Late delivery   
2        -247.779999          309.720001  Shipping on time   
3          22.860001          304.809998  Advance shipping   
4         134.210007          298.250000  Advance shipping   

   Late_delivery_risk  Category Id   Category Name Customer City  \
0                   0           73  Sporting Goods        Caguas   
1                   1           73  Sporting Goods        Caguas   
2                   0           73  Sporting Goods      San Jose   
3                   0           73  Sporting Goods   Los Angeles   
4                   0           73  Sporting Goods        Caguas   

  Customer Country Customer Email Customer Fname  Customer Id Customer Lname  \
0      Puerto Rico      XXXXXXXXX          Cally        20755       Holloway   
1      Puerto Rico      XXXXXXXXX          Irene        19492           Luna   
2          EE. UU.      XXXXXXXXX        Gillian        19491      Maldonado   
3          EE. UU.      XXXXXXXXX           Tana        19490           Tate   
4      Puerto Rico      XXXXXXXXX           Orli        19489      Hendricks   

  Customer Password Customer Segment Customer State           Customer Street  \
0         XXXXXXXXX         Consumer             PR  5365 Noble Nectar Island   
1         XXXXXXXXX         Consumer             PR          2679 Rustic Loop   
2         XXXXXXXXX         Consumer             CA      8510 Round Bear Gate   
3         XXXXXXXXX      Home Office             CA           3200 Amber Bend   
4         XXXXXXXXX        Corporate             PR  8671 Iron Anchor Corners   

   Customer Zipcode  Department Id Department Name   Latitude   Longitude  \
0             725.0              2         Fitness  18.251453  -66.037056   
1             725.0              2         Fitness  18.279451  -66.037064   
2           95125.0              2         Fitness  37.292233 -121.881279   
3           90027.0              2         Fitness  34.125946 -118.291016   
4             725.0              2         Fitness  18.253769  -66.037048   

         Market  Order City Order Country  Order Customer Id  \
0  Pacific Asia      Bekasi     Indonesia              20755   
1  Pacific Asia     Bikaner         India              19492   
2  Pacific Asia     Bikaner         India              19491   
3  Pacific Asia  Townsville     Australia              19490   
4  Pacific Asia  Townsville     Australia              19489   

  order date (DateOrders)  Order Id  Order Item Cardprod Id  \
0         1/31/2018 22:56     77202                    1360   
1         1/13/2018 12:27     75939                    1360   
2         1/13/2018 12:06     75938                    1360   
3         1/13/2018 11:45     75937                    1360   
4         1/13/2018 11:24     75936                    1360   

   Order Item Discount  Order Item Discount Rate  Order Item Id  \
0            13.110000                      0.04         180517   
1            16.389999                      0.05         179254   
2            18.030001                      0.06         179253   
3            22.940001                      0.07         179252   
4            29.500000                      0.09         179251   

   Order Item Product Price  Order Item Profit Ratio  Order Item Quantity  \
0                    327.75                     0.29                    1   
1                    327.75             

In [5]:
print(df.shape)

(180519, 53)


From the output, we can see that the table contains ***180519*** rows and ***53*** columns.

#### Now let’s try printing out the column names using: ***columns***

In [6]:
print(df.columns)

Index(['Type', 'Days for shipping (real)', 'Days for shipment (scheduled)',
       'Benefit per order', 'Sales per customer', 'Delivery Status',
       'Late_delivery_risk', 'Category Id', 'Category Name', 'Customer City',
       'Customer Country', 'Customer Email', 'Customer Fname', 'Customer Id',
       'Customer Lname', 'Customer Password', 'Customer Segment',
       'Customer State', 'Customer Street', 'Customer Zipcode',
       'Department Id', 'Department Name', 'Latitude', 'Longitude', 'Market',
       'Order City', 'Order Country', 'Order Customer Id',
       'order date (DateOrders)', 'Order Id', 'Order Item Cardprod Id',
       'Order Item Discount', 'Order Item Discount Rate', 'Order Item Id',
       'Order Item Product Price', 'Order Item Profit Ratio',
       'Order Item Quantity', 'Sales', 'Order Item Total',
       'Order Profit Per Order', 'Order Region', 'Order State', 'Order Status',
       'Order Zipcode', 'Product Card Id', 'Product Category Id',
       'Product De

#### We can use the ***info()*** method to output some general information about the dataframe:

In [7]:
print(df.info()) # Check the data information like which type of data the column contains

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180519 entries, 0 to 180518
Data columns (total 53 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   Type                           180519 non-null  object 
 1   Days for shipping (real)       180519 non-null  int64  
 2   Days for shipment (scheduled)  180519 non-null  int64  
 3   Benefit per order              180519 non-null  float64
 4   Sales per customer             180519 non-null  float64
 5   Delivery Status                180519 non-null  object 
 6   Late_delivery_risk             180519 non-null  int64  
 7   Category Id                    180519 non-null  int64  
 8   Category Name                  180519 non-null  object 
 9   Customer City                  180519 non-null  object 
 10  Customer Country               180519 non-null  object 
 11  Customer Email                 180519 non-null  object 
 12  Customer Fname                

# *Data Cleaning*

From the output of ***info*** we can see that ***shipping date (DateOrders)*** and ***order date (DateOrders)*** is showing ***object*** data type. It should be ***datetime*** data type. So, Let's make it

In [8]:
df['shipping date (DateOrders)'] = pd.to_datetime(df['shipping date (DateOrders)'], format='%m/%d/%Y %H:%M')
df['order date (DateOrders)'] = pd.to_datetime(df['order date (DateOrders)'], format='%m/%d/%Y %H:%M')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180519 entries, 0 to 180518
Data columns (total 53 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   Type                           180519 non-null  object        
 1   Days for shipping (real)       180519 non-null  int64         
 2   Days for shipment (scheduled)  180519 non-null  int64         
 3   Benefit per order              180519 non-null  float64       
 4   Sales per customer             180519 non-null  float64       
 5   Delivery Status                180519 non-null  object        
 6   Late_delivery_risk             180519 non-null  int64         
 7   Category Id                    180519 non-null  int64         
 8   Category Name                  180519 non-null  object        
 9   Customer City                  180519 non-null  object        
 10  Customer Country               180519 non-null  object        
 11  

Now ***shipping date (DateOrders)*** and ***order date (DateOrders)*** columns are showing correct data type.

In [9]:
# UDF for getting duplicate columns name
def getDuplicateColumns(df):
    duplicateColumnNames = set() # Define empty set
     
    for x in range(df.shape[1]): # Iterate through all the columns of DataFrame
        col = df.iloc[:, x]
        for y in range(x + 1, df.shape[1]): #Iterate through all the columns in DataFrame from (x + 1)th index to last index
            otherCol = df.iloc[:, y]
             
            # If two columns are equal then adding to the set
            if col.equals(otherCol):
                duplicateColumnNames.add(df.columns.values[y])
                 
    # Return list of unique column names whose contents are duplicates
    return list(duplicateColumnNames)

columns = getDuplicateColumns(df)
print(columns)
print("\nTotal duplicate columns: ", len(columns))

['Product Card Id', 'Customer Password', 'Product Price', 'Order Customer Id', 'Product Category Id', 'Order Profit Per Order', 'Order Item Total']

Total duplicate columns:  7


These columns are duplicate, So we can drop these colums. Let's drop those

In [10]:
df = df.drop(columns = getDuplicateColumns(df), inplace=False)
print(df.shape)

(180519, 46)


Earlier there was 53 columns, After removing ***7*** duplicate columns, Now it have 46 columns

In [11]:
# UDF for getting duplicate columns name
def getOneDistinctValueCols(df):
    duplicateColumnNames = set() # Define empty set
    uniqueColumnNames = {}
    
    for col in list(df.columns): # Iterate through all the columns of DataFrame
        listUniqueValue = df[col].unique()
        if (len(listUniqueValue) == 1): # If column contains only signle unqiue value then adding to the dictionary
            uniqueColumnNames[col] = listUniqueValue[0];
    
    # Return dict of signle unique column names along with values.
    return dict(uniqueColumnNames)

columns = getOneDistinctValueCols(df)
print(columns)
print("\nTotal columns: ", len(columns))

{'Customer Email': 'XXXXXXXXX', 'Product Description': nan, 'Product Status': 0}

Total columns:  3


From the above result we can see that ***3*** columns contains only one unique values. So, we can drop those columns. Let's do it

In [12]:
df = df.drop(columns = getOneDistinctValueCols(df).keys(), inplace=False)
print(df.shape)

(180519, 43)


Earlier there was 53 columns, After removing ***3*** duplicate columns, Now it have 43 columns

In [13]:
# Merge customer first name & last name in single column 
df["Customer Name"] = df["Customer Fname"].astype(str) + " " + df["Customer Lname"].astype(str)

In [14]:
df_a = df.drop(columns = ["Customer Fname", "Customer Lname"], inplace=False) # Drop Customer Fname & Customer Lname columns
print(df_a.shape)

(180519, 42)


### Now let's check whether there have any null values or not

In [15]:
print(df_a.isnull().sum())

Type                                  0
Days for shipping (real)              0
Days for shipment (scheduled)         0
Benefit per order                     0
Sales per customer                    0
Delivery Status                       0
Late_delivery_risk                    0
Category Id                           0
Category Name                         0
Customer City                         0
Customer Country                      0
Customer Id                           0
Customer Segment                      0
Customer State                        0
Customer Street                       0
Customer Zipcode                      3
Department Id                         0
Department Name                       0
Latitude                              0
Longitude                             0
Market                                0
Order City                            0
Order Country                         0
order date (DateOrders)               0
Order Id                              0


From the result, We can see there is ***Order Zipcode*** column contains null value. And if we think about this column it is irrelevant in data analysis. So, we can drop this column as well. 

In [16]:
df_b = df_a.drop(columns = ["Order Zipcode"], inplace=False)
print(df_b.isnull().sum())

Type                             0
Days for shipping (real)         0
Days for shipment (scheduled)    0
Benefit per order                0
Sales per customer               0
Delivery Status                  0
Late_delivery_risk               0
Category Id                      0
Category Name                    0
Customer City                    0
Customer Country                 0
Customer Id                      0
Customer Segment                 0
Customer State                   0
Customer Street                  0
Customer Zipcode                 3
Department Id                    0
Department Name                  0
Latitude                         0
Longitude                        0
Market                           0
Order City                       0
Order Country                    0
order date (DateOrders)          0
Order Id                         0
Order Item Cardprod Id           0
Order Item Discount              0
Order Item Discount Rate         0
Order Item Id       

Now, we can see there is ***no null values*** in the above columns

# *Univarite Analysis*

* Univariate analysis:- provides summary statistics for each field in the raw data set (or) summary only on one variable. Ex:- CDF,PDF,Box plot, Violin plot.(don't worry, will see below what each of them is)
* Bivariate analysis:- is performed to find the relationship between each variable in the dataset and the target variable of interest (or) using 2 variables and finding the relationship between them.Ex:-Box plot, Violin plot.
* Multivariate analysis:- is performed to understand interactions between different fields in the dataset (or) finding interactions between variables more than 2. Ex:- Pair plot and 3D scatter plot.